File: api/zornigor/db/stories.py
from datetime import datetime
from sqlite3 import IntegrityError
from typing import List, Optional, Union

from sqlalchemy import and_, insert, select, update, delete

from zornigor.db.errors import InvalidState
from zornigor.db.models import NewStory, Story, UpdateStory
from zornigor.db.projects import get_project_last_story_id, update_project_last_story_id
from zornigor.db.tables import stories
from zornigor.db.utility import get_database


async def create_story(story: Union[Story, NewStory]):
    db = get_database(strict=True)  # type: ignore
    async with db.transaction():
        values = story.dict()

        if type(story) == NewStory:
            # use incremental story id by project
            last_story_id = await get_project_last_story_id(story.project)
            story_id = last_story_id + 1

            await update_project_last_story_id(story.project, story_id)
            values["id"] = story_id

        stmt = insert(stories).values(**values)
        try:
            result = await db.execute(stmt)
        except IntegrityError as exc:
            raise InvalidState(
                f"Invalid state '{story.state}' for story in project {story.project}"
            ) from exc

        return result


async def get_story(project: str, id: int) -> Optional[Story]:
    db = get_database(strict=True)  # type: ignore
    stmt = select(
        stories.c.id,
        stories.c.title,
        stories.c.description,
        stories.c.created,
        stories.c.modified,
        stories.c.project,
        stories.c.state,
    ).where(and_(stories.c.project == project, stories.c.id == id))

    result = await db.fetch_one(stmt)
    if result is None:
        return None

    return Story(
        id=result.id,
        title=result.title,
        description=result.description,
        created=result.created,
        modified=result.modified,
        project=result.project,
        state=result.state,
    )


async def list_stories(project: str) -> List[Story]:
    db = get_database(strict=True)  # type: ignore
    stmt = select(
        stories.c.id,
        stories.c.title,
        stories.c.description,
        stories.c.created,
        stories.c.modified,
        stories.c.project,
        stories.c.state,
    ).where(stories.c.project == project)
    return [
        Story(
            id=row.id,
            title=row.title,
            description=row.description,
            created=row.created,
            modified=row.modified,
            project=row.project,
            state=row.state,
        )
        async for row in db.iterate(stmt)
    ]


async def update_story(project: str, story_id: int, story: UpdateStory):
    db = get_database(strict=True)  # type: ignore
    values = story.dict()
    values["modified"] = datetime.now()
    stmt = update(stories).where(and_(stories.c.project == project, stories.c.id == story_id)).values(**values)
    await db.execute(stmt)


async def delete_story(project: str, story_id: int):
    db = get_database(strict=True)  # type: ignore
    stmt = delete(stories).where(and_(stories.c.project == project, stories.c.id == story_id))
    await db.execute(stmt)

File: frontend/web/src/api/stories.js
import axios from "axios"

import * as  validators from "@/api/validators"

import { BASE_URL } from "@/globals"


function create_story(project_id, story) {
    return axios
        .post(BASE_URL + "/project/" + project_id + "/stories", story)
        .then((response) => {
            if (response.status != 201) {
                throw Error(`API error ${response.status}: ${response.data}`);
            }
        })
        .catch((error) => {
            console.log(`Error while POST /project/${project_id}/stories: ${error.message}`);
            throw error;
        })
}

function get_story(project_id, story_id) {
    return axios
        .get(BASE_URL + "/project/" + project_id + "/story/" + story_id)
        .then((response) => {
            if (response.status != 200) {
                throw Error(`API error ${response.status}: ${response.data}`);
            }

            if (!validators.is_valid_story(response.data)) {
                let api_error = validators.is_valid_story.errors[-1];
                throw new Error(`Invalid API data: ${api_error}`);
            }

            let story = response.data;
            return story;
        })
        .catch((error) => {
            console.log(`Error while GET /project/${project_id}/story/${story_id}: ${error.message}`);
            throw error;
        })
}

function list_stories(project_id) {
    return axios
        .get(BASE_URL + "/project/" + project_id + "/stories")
        .then((response) => {
            if (response.status != 200) {
                throw Error(`API error ${response.status}: ${response.data}`);
            }

            if (!validators.is_valid_story_list(response.data)) {
                let api_error = validators.is_valid_story_list.errors[-1];
                throw new Error(`Invalid API data: ${api_error}`);
            }

            let stories = response.data;
            return stories;
        })
        .catch((error) => {
            console.log(`Error while GET /project/${project_id}/stories: ${error.message}`);
            throw error;
        })
}


function update_story(project_id, story_id, update) {
    return axios
        .patch(BASE_URL + "/project/" + project_id + "/story/" + story_id, update)
        .then((response) => {
            if (response.status != 204) {
                throw Error(`API error ${response.status}: ${response.data}`);
            }
        })
        .catch((error) => {
            console.log(`Error while PATCH /project/${project_id}/story/${story_id}: ${error.message}`);
            throw error;
        })
}


export {
    create_story,
    get_story,
    list_stories,
    update_story,
}
